File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        tmp = set(dictionary)
        n = len(s)
        dp = [inf] * (n + 1)
        dp[0] = 0
        for i in range(n):
            dp[i+1] = min(dp[i] + 1, dp[i+1])
            for j in range(i, n):
                if s[i:j+1] in tmp:
                    dp[j+1] = min(dp[j+1], dp[i])
        return dp[-1]

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(string s, vector<string>& arr) {
        unordered_set<string> st(arr.begin(), arr.end());
        int n = s.length();

        vector<int> dp(n + 1);
        for (int i = 1; i <= n; i++) {
            dp[i] = dp[i - 1] + 1;

            for (int j = i; j >= 1; j--) {
                string sub = s.substr(j - 1, i - (j - 1));

                if (st.count(sub) > 0) {
                    dp[i] = min(dp[i], dp[j - 1]);
                }
            }
        }

        return dp[n];
    }
};